table/transaction.go

```diff
@@ -0,0 +1,346 @@
+	out := true
+	for _, res := range v {
+		out = res() && out
+	}
+
+	return out
+}
+
+func (sc *schemaCompatVisitor) Field(n iceberg.NestedField, v func() bool) bool {
+	return sc.isFieldCompat(n) && v()
+}
+
+func (sc *schemaCompatVisitor) List(l iceberg.ListType, v func() bool) bool {
+	return sc.isFieldCompat(l.ElementField()) && v()
+}
+
+func (sc *schemaCompatVisitor) Map(m iceberg.MapType, vk, vv func() bool) bool {
+	return sc.isFieldCompat(m.KeyField()) && sc.isFieldCompat(m.ValueField()) && vk() && vv()
+}
+
+func (sc *schemaCompatVisitor) Primitive(p iceberg.PrimitiveType) bool {
+	return true
+}
+
+type snapshotUpdate struct {
+	txn           *Transaction
+	io            io.WriteFileIO
+	snapshotProps iceberg.Properties
+}
+
+func (s snapshotUpdate) fastAppend() *snapshotProducer {
+	return newFastAppendFilesProducer(OpAppend, s.txn, s.io, nil, s.snapshotProps)
+}
+
+func (s snapshotUpdate) mergeAppend() *snapshotProducer {
+	return newMergeAppendFilesProducer(OpAppend, s.txn, s.io, nil, s.snapshotProps)
+}
+
+type Transaction struct {
+	tbl  *Table
+	meta *MetadataBuilder
+
+	reqs []Requirement
+
+	mx        sync.Mutex
+	committed bool
+}
+
+func (t *Transaction) apply(updates []Update, reqs []Requirement) error {
+	t.mx.Lock()
+	defer t.mx.Unlock()
+
+	if t.committed {
+		return errors.New("transaction has already been committed")
+	}
+
+	current, err := t.meta.Build()
+	if err != nil {
+		return err
+	}
+
+	for _, r := range reqs {
+		if err := r.Validate(current); err != nil {
+			return err
+		}
+	}
+
+	existing := map[string]struct{}{}
+	for _, r := range t.reqs {
+		existing[r.GetType()] = struct{}{}
+	}
+
+	for _, r := range reqs {
+		if _, ok := existing[r.GetType()]; !ok {
+			t.reqs = append(t.reqs, r)
+		}
+	}
+
+	prevUpdates, prevLastUpdated := len(t.meta.updates), t.meta.lastUpdatedMS
+	for _, u := range updates {
+		if err := u.Apply(t.meta); err != nil {
+			return err
+		}
+	}
+
+	if prevUpdates < len(t.meta.updates) {
+		if prevLastUpdated == t.meta.lastUpdatedMS {
+			t.meta.lastUpdatedMS = time.Now().UnixMilli()
+		}
+	}
+
+	return nil
+}
+
+func (t *Transaction) updateSnapshot(props iceberg.Properties) snapshotUpdate {
+	return snapshotUpdate{
+		txn:           t,
+		io:            t.tbl.fs.(io.WriteFileIO),
+		snapshotProps: props,
+	}
+}
+
+func (t *Transaction) appendSnapshotProducer(props iceberg.Properties) *snapshotProducer {
+	manifestMerge := t.meta.props.GetBool(ManifestMergeEnabledKey, ManifestMergeEnabledDefault)
+
+	updateSnapshot := t.updateSnapshot(props)
+	if manifestMerge {
+		return updateSnapshot.mergeAppend()
+	}
+
+	return updateSnapshot.fastAppend()
+}
+
+func (t *Transaction) SetProperties(props iceberg.Properties) error {
+	if len(props) > 0 {
+		return t.apply([]Update{NewSetPropertiesUpdate(props)}, nil)
+	}
+
+	return nil
+}
+
+func (t *Transaction) Append(rdr array.RecordReader, snapshotProps iceberg.Properties) error {
```

Comment:
This doesn't anything, right?

Reply:
correct, currently it doesn't do anything yet. I'll change it to return a not implemented error